### PR TITLE
virtual service supports traffic mirroring to multiple destinations

### PR DESCRIPTION
**Pull request by fatedier**
Fix: https://github.com/istio/istio/issues/13330

### CODE REVIEW

**Review by howardjohn — COMMENTED** [review]
Mostly lgtm. I want to check gateway-api plans though: (https://kubernetes.slack.com/archives/CR0H13KGA/p1686580503265289) for alignment

**Comment by howardjohn** [comment]
Will be good to resolve https://github.com/kubernetes-sigs/gateway-api/issues/2109 first

File: networking/v1beta1/virtual_service.proto [hunk 1]
```diff
@@ -633,6 +633,14 @@ message HTTPRoute {
   // destination.
   Destination mirror = 9;
 
+  // Mirror HTTP traffic to another destinations in addition to forwarding
```
Comment (linsun) [inline networking/v1beta1/virtual_service.proto:636]: another destination or destinations?

**Comment by linsun** [comment]
Is there an implementation for this? If not, should we hide the API docs first?

**Comment by fatedier** [comment]
> Is there an implementation for this? If not, should we hide the API docs first?

I am not very familiar with the process. First, I would like to confirm if modifying this API is acceptable, and then proceed with its implementation.

**Comment by hzxuzhonghu** [comment]
lin means there should be a following implement in istio, otherwise mark it hidden

**Comment by fatedier** [comment]
Hide from docs done.

**Comment by fatedier** [comment]
@howardjohn Hi, I found that https://github.com/kubernetes-sigs/gateway-api/issues/2109 has been accepted. Can the modifications to this API continue to progress?

**Comment by howardjohn** [comment]
Looks like GW is going to approve, just waiting on official signoff then will approve this. Thanks for the patience

**Review by howardjohn — APPROVED** [review]

File: networking/v1alpha3/virtual_service.proto [hunk 2]
```diff
@@ -633,6 +633,17 @@ message HTTPRoute {
   // destination.
   Destination mirror = 9;
 
+  // $hide_from_docs
+  // Hide this from doc until implementing this.
```
Comment (ericvn) [inline networking/v1alpha3/virtual_service.proto:637]:
```suggestion
  // Hide this from doc until implemented.
```

**Comment by matined** [comment]
May I ask if you know how much time after the merge we'll have to wait for this feature to be included in any release?

**Comment by fatedier** [comment]
> May I ask if you know how much time after the merge we'll have to wait for this feature to be included in any Istio release?

There is already a draft PR in istio repo https://github.com/istio/istio/pull/45944. I believe that after the modification of the API is merged, the implementation of this feature will not take too long.

File: networking/v1alpha3/virtual_service.proto [hunk 3]
```diff
@@ -633,6 +633,17 @@ message HTTPRoute {
...
+  //
+  // Mirror HTTP traffic to another destinations in addition to forwarding
+  // the requests to the intended destination. Mirrored traffic is on a
+  // best effort basis where the sidecar/gateway will not wait for the
+  // mirrored cluster to respond before returning the response from the
+  // original destination. Statistics will be generated for the mirrored
+  // destination.
```
[the remark this answers is not in the document]
Comment (ericvn) [reply networking/v1alpha3/virtual_service.proto:644]: Le me tweak this some more since I see you are also trying to mirror what we already had.

**Review by ericvn — APPROVED** [review]

**Pull request merged**